File: zone-server/src/main/kotlin/net/bestia/zone/world/prop/interact/InteractEntityCMSG.kt
package net.bestia.zone.world.prop.interact

import net.bestia.bnet.proto.InteractEntityCMSGProto
import net.bestia.zone.message.CMSG
import net.bestia.zone.script.ScriptArgs
import net.bestia.zone.util.EntityId

/**
 * The player clicked on something in the world. What that means is decided by what the target is - see
 * [InteractEntityHandler].
 */
data class InteractEntityCMSG(
  override val playerId: Long,
  val targetEntityId: EntityId,
  val args: ScriptArgs
) : CMSG {

  companion object {
    fun fromBnet(playerId: Long, bnet: InteractEntityCMSGProto.InteractEntityCMSG): InteractEntityCMSG {
      return InteractEntityCMSG(
        playerId = playerId,
        targetEntityId = bnet.entityId,
        args = ScriptArgs.fromBnet(bnet.args)
      )
    }
  }
}

File: bestia-client/src/Bnet/Message/Map/InteractEntityCMSG.cs
using Bnet;

namespace BestiaBehemothClient.Bnet.Message.Map
{
  /// <summary>
  /// "I clicked on that." What the click means is decided server-side, from what the target is.
  /// </summary>
  /// <remarks>
  /// The general form of a world interaction, where <see cref="CollectPropCMSG"/> and AttackEntityCMSG are
  /// the specific ones. Ids here are live ECS ids and only valid while the client holds the thing - sending a
  /// stale one is harmless, since the server cannot resolve it and refuses.
  /// </remarks>
  public partial class InteractEntityCMSG : ICMSG
  {
    public long EntityId { get; set; }

    /// <summary>Whatever the interaction needed the player to choose first. Null for a plain click.</summary>
    public Message.ScriptArgs Args { get; set; }

    public override Envelope ToEnvelope()
    {
      var interact = new global::Bnet.InteractEntityCMSG
      {
        EntityId = (ulong)EntityId,
        Args = Args?.ToProto()
      };

      return new Envelope
      {
        InteractEntity = interact
      };
    }
  }
}
